### PR TITLE
List types to bind in Python explicitly

### PR DESCRIPTION
**Pull request by SimonHeybrock**

### CODE REVIEW

**Review by jl-wynen — APPROVED** [review]
Looks good.
But please update `reference/dtype.html`. It says
> You can find a full list in the docs of the [scipp.DType](https://scipp.github.io/generated/classes/scipp.DType.html) class. But note that many of those dtypes are only meant for internal use.

**Review by jl-wynen — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
@jl-wynen I think I got this working now, please have another look.

File: lib/python/dtype.cpp [hunk 1]
```diff
@@ -95,7 +90,10 @@ objects containing binned data. They cannot be used directly to create arrays of
            dtype<core::bin<Dataset>>,
            dtype<python::PyObject>,
        }) {
-    std::cout << i++ << '\n';
+    // For some reason OSX causes trouble when we return `t` directly, with the
+    // dtype not matching the one in the registry (possibly related to problems
+    // of symbols across library boundaries). Instead, we find by name and
+    // return the key.
```
Comment (jl-wynen) [inline lib/python/dtype.cpp:96]: The old implementation used a lambda that captured `t` by ref and stored it in pybind11's storage. So the ref outlived the initializer list in the for loop above. Maybe on OSX, that list contains local copies but on Ubuntu and Windows it contains refs to the static `dtype` objects?
Did you try capturing `t` by value?

Comment (SimonHeybrock) [reply lib/python/dtype.cpp:96]: Hmm, good point, I will try.

**Comment by SimonHeybrock** [comment]
@jl-wynen You were right! This worked.

**Review by jl-wynen — APPROVED** [review]
Nice :smiley:

**Pull request merged**